vmware: improve solidfire storage plugin integration and fix cases

rohityadavcloud opened the pull request:
This fixes the following cases in which Solidfire storage integration caused issues when using Solidfire datadisks with VMware:

1. Take Volume Snapshot of Solidfire data disk
2. Delete an active Instance with Solidfire data disk attached
3. Attach used existing Solidfire data disk to a running/stopped VM
4. Stop and Start an instance with Solidfire data disks attached
5. Expand disk by resizing Solidfire data disk by providing size
6. Expand disk by changing disk offering for the Solidfire data disk

Function | Operation | Description | Fix Status
-- | -- | -- | --
Create | Take Volume Snapshot | Take data disk volume snapshot on active instance | :white_check_mark:
Delete | Delete an active Instance | Delete an active Instance with attached 1:1 data disks | :white_check_mark:
Update | Attach used existing data disk | Attach a previous used data volume to a instance | :white_check_mark:
Update | Stop and Start a Instance | Stop and Start a Instance with attached 1:1 data disks | :white_check_mark:
Update | Resize Data volume | Expand data disk on demand | :white_check_mark:
Update | Change disk offering for the data volume | Expand data disk changing disk offering | :white_check_mark:
Update | Migrate data volume | Migrate a data volume between primaries storages Solidfire based | Unsupported feature

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [x] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Tested in a ACS 4.17 adv zone env with VMware and Solidfire setup. Tested all the following cases:

<img width="1231" alt="Screenshot 2023-07-21 at 1 02 29 PM" src="https://github.com/apache/cloudstack/assets/95203/39ce7098-7d75-44f7-8086-1116bb4c77f3">

rohityadavcloud commented on plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/manager/VmwareStorageManagerImpl.java:
```diff
@@ -1113,6 +1113,9 @@ private long getVMSnapshotChainSize(VmwareContext context, VmwareHypervisorHost
         DatastoreMO dsMo = new DatastoreMO(context, morDs);
         HostDatastoreBrowserMO browserMo = dsMo.getHostDatastoreBrowserMO();
         String datastorePath = (new DatastoreFile(dsMo.getName(), vmName)).getPath();
+        if (datastorePath.startsWith("[-iqn.")) { // ex. [-iqn.2010-01.com.solidfire:3p53.data-9999.97-0] i-2-9999-VM
```
Note: similar code exists in the caller that checks for the specific datastore string

rohityadavcloud replied:
I'll attempt refactoring next week

rohityadavcloud commented on plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -927,6 +927,7 @@ private Answer execute(ResizeVolumeCommand cmd) {
                 ClusterMO clusterMO = new ClusterMO(context, morCluster);
 
                 List<Pair<ManagedObjectReference, String>> lstHosts = clusterMO.getClusterHosts();
+                _storageProcessor.rescanAllHosts(context, lstHosts, true, true);
```
There's some issue with VMware not able to detect resized/modification in the LUN size, this is a workaround for VMware to see an updated/resized LUN. Otherwise VMFS datastore doesn't see it can resize and silently fails.

shwstppr replied:
I hope this won't cause an issue when there are many hosts in one cluster. Not sure if there is a way to limit rescanning to specific hosts or should we limit it to Solidfire volumes?

rohityadavcloud replied:
I think we'll need to rescan all the hosts in the cluster; yes that could make the process very slow. If necessary I can add in some logic to scan only when datastore is iscsi (start with -iqn...)

rohityadavcloud replied:
@shwstppr I had a code check, line no. 925 ensures this is run only in case of managed storage such as solidfire; so wouldn't affect non-managed storage such as nfs primary storage.

rohityadavcloud replied:
I'll still add a check

rohityadavcloud commented on engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:
```diff
@@ -1879,8 +1879,8 @@ public void prepare(VirtualMachineProfile vm, DeployDestination dest) throws Sto
-                        // This might impact other managed storages, grant access for PowerFlex storage pool only
-                        if (pool.getPoolType() == Storage.StoragePoolType.PowerFlex) {
+                        // This might impact other managed storages, grant access for PowerFlex and Iscsi/Solidfire storage pool only
+                        if (pool.getPoolType() == Storage.StoragePoolType.PowerFlex || pool.getPoolType() == Storage.StoragePoolType.Iscsi) {
```
This is to workaround the issue mentioned in the solidfire plugin, to grant access (that LUN is accessible to vmware).

rohityadavcloud commented on plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:
```diff
@@ -2826,6 +2826,9 @@ private ManagedObjectReference getVmfsDatastore(VmwareContext context, VmwareHyp
             morDs = firstHostDatastoreSystemMO.createVmfsDatastore(datastoreName, hostScsiDisk);
+        } else {
+            // in case of iSCSI/solidfire 1:1 VMFS datastore could be inaccessible
+            mountVmfsDatastore(new DatastoreMO(context, morDs), lstHosts);
```
Workaround this checks and mounts inaccessible datastore. For others it will simply skip, for solidfire it will check and mount the datastore on the hosts.

rohityadavcloud commented on plugins/storage/volume/solidfire/src/main/java/org/apache/cloudstack/storage/datastore/driver/SolidFirePrimaryDataStoreDriver.java:
```diff
@@ -188,12 +192,25 @@ public boolean grantAccess(DataObject dataObject, Host host, DataStore dataStore
+        // Workaround: don't unplug iscsi lun when volume is attached to a VM
+        // This is regression workaround from upper layers which are calling
+        // ::releaseVmResources that calls the revoke on attached disk of a VM
+        if (dataObject.getType() == DataObjectType.VOLUME) {
```
Explained the workaround in the comments, prior to ScaleIO integration the revokeAccess() wasn't called by upper layers say when a VM is stopped. We don't want to eject/remove the LUN from VMware unless the volume is detached or the VM/volume is to be deleted.

shwstppr replied:
can be a separate method

rohityadavcloud replied:
Can be done

rohityadavcloud replied:
done

weizhouapache commented:
great work @rohityadavcloud

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

DaanHoogland reviewed (approved):
clgtm, but i cannot test this. Can you extract the workarounds (and containg blocks) in separate methods and where feasible add unit tests?

shwstppr reviewed (approved):
Code change looks good in terms of logic. Left some comments. May need manual testing with different storage providers

blueorangutan commented:
Packaging result [SF]: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 6532

rohityadavcloud commented:
@blueorangutan test matrix

rohityadavcloud commented:
@blueorangutan test matrix

blueorangutan commented:
@rohityadavcloud a [SF] Trillian-Jenkins matrix job (centos7 mgmt + xenserver71, rocky8 mgmt + vmware67u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-7135)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43125 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7135-xenserver-71.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_cancel_host_maintenace_with_no_migration_jobs | `Error` | 309.66 | test_host_maintenance.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-7137)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 45954 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7137-kvm-centos7.zip
Smoke tests completed. 106 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_migrate_VM_and_root_volume | `Error` | 76.80 | test_vm_life_cycle.py
test_02_migrate_VM_with_two_data_disks | `Error` | 53.63 | test_vm_life_cycle.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 303.84 | test_hostha_kvm.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-7136)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 51697 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7136-vmware-67u3.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_deploy_vm_on_specific_host | `Error` | 18.77 | test_vm_deployment_planner.py
test_02_deploy_vm_on_specific_cluster | `Error` | 3602.92 | test_vm_deployment_planner.py
test_03_deploy_vm_on_specific_pod | `Error` | 4.42 | test_vm_deployment_planner.py
test_04_deploy_vm_on_host_override_pod_and_cluster | `Error` | 4.45 | test_vm_deployment_planner.py
test_05_deploy_vm_on_cluster_override_pod | `Error` | 2.33 | test_vm_deployment_planner.py

harikrishna-patnala reviewed (approved):
Code LGTM, all checks and accessories changes make sense for fixing the issues. I did not test it though.

rohityadavcloud commented:
Thanks @shwstppr @DaanHoogland @weizhouapache @harikrishna-patnala I've incorporated your review and refactored places that wouldn't entagle the wires too much and added unit tests for cases. Rest I've re-tested all cases again:
<img width="1233" alt="Screenshot 2023-07-25 at 1 17 09 PM" src="https://github.com/apache/cloudstack/assets/95203/63c56bc1-7653-4ffe-91ec-bc222c46c0c8">

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_multiplication_x: debian :heavy_check_mark: suse15. SL-JID 6557

rohityadavcloud commented:
@blueorangutan test matrix

blueorangutan commented:
@rohityadavcloud a [SF] Trillian-Jenkins matrix job (centos7 mgmt + xenserver71, rocky8 mgmt + vmware67u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-7161)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 40195 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7161-xenserver-71.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_cancel_host_maintenace_with_no_migration_jobs | `Error` | 307.18 | test_host_maintenance.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 307.17 | test_host_maintenance.py
test_03_cancel_host_maintenace_with_migration_jobs_failure | `Error` | 0.22 | test_host_maintenance.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-7163)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 40683 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7163-kvm-centos7.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_migrate_VM_and_root_volume | `Error` | 75.75 | test_vm_life_cycle.py
test_02_migrate_VM_with_two_data_disks | `Error` | 51.38 | test_vm_life_cycle.py

rohityadavcloud commented:
@blueorangutan test matrix

rohityadavcloud commented:
@blueorangutan test matrix

blueorangutan commented:
@rohityadavcloud a [SF] Trillian-Jenkins matrix job (centos7 mgmt + xenserver71, rocky8 mgmt + vmware67u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-7172)<b/>

blueorangutan commented:
<b>[SF] Trillian test result (tid-7171)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 37558 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7171-xenserver-71.zip
Smoke tests completed. 108 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>[SF] Trillian test result (tid-7173)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 48780 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7173-kvm-centos7.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_migrate_VM_and_root_volume | `Error` | 76.57 | test_vm_life_cycle.py
test_02_migrate_VM_with_two_data_disks | `Error` | 55.40 | test_vm_life_cycle.py

weizhouapache commented:
@rohityadavcloud 
is this ready for merge ?

rohityadavcloud commented:
Yes, @weizhouapache but @NuxRo needs to sign off with his testing. I'll change the PR to draft.

weizhouapache commented:
> Yes, @weizhouapache but @NuxRo needs to sign off with his testing. I'll change the PR to draft.

Ok thanks @rohityadavcloud

weizhouapache commented:
cc @NuxRo 
plaese kindly approve it if it is tested good

weizhouapache reviewed (approved):
code lgtm

rohityadavcloud commented:
@NuxRo completed QA on this, against a community-supplied 4.17 Solidfire env (before he went on holidays). Now with @andrijapanicsb to finalise review. Thanks.

@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 6877

blueorangutan commented:
@weizhouapache a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-67u3) has been kicked to run smoke tests

blueorangutan commented:
@weizhouapache [SF] unsupported parameters provided. Supported mgmt server os are: `centos7, centos6, suse15, alma8, ubuntu18, ubuntu22, ubuntu20, rocky8, alma9`. Supported hypervisors are: `kvm-centos6, kvm-centos7, kvm-rocky8, kvm-alma8, kvm-alma9, kvm-ubuntu18, kvm-ubuntu20, kvm-ubuntu22, kvm-suse15, vmware-55u3, vmware-60u2, vmware-65u2, vmware-67u3, vmware-70u1, vmware-70u2, vmware-70u3, vmware-80, vmware-80u1, xenserver-65sp1, xenserver-71, xenserver-74, xcpng74, xcpng76, xcpng80, xcpng81, xcpng82`

weizhouapache commented:
@blueorangutan test rocky8 vmware-70u3

blueorangutan commented:
@weizhouapache a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-70u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-7531)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 49474 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7531-vmware-67u3.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_upgrade_kubernetes_cluster | `Failure` | 560.83 | test_kubernetes_clusters.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-7532)</b>
Environment: vmware-70u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 47727 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7761-t7532-vmware-70u3.zip
Smoke tests completed. 107 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_upgrade_kubernetes_cluster | `Failure` | 585.85 | test_kubernetes_clusters.py
